### PR TITLE
Improve test to not rely on thread slowness

### PR DESCRIPTION
**Pull request by s1monw**
We have to swap the second latch before we count it down otherwise
threads might be faster than the test. This has happend on a recent
CI failure: https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+master+multijob-os-compatibility/os=ubuntu/121/console

This commit also adds a synchronized on the close method since it's
canceling and modifying a member varialbe that is assigned under lock.

### CODE REVIEW

File: core/src/test/java/org/elasticsearch/index/IndexServiceTests.java [hunk 1]
```diff
@@ -27,6 +27,7 @@
 import org.elasticsearch.common.compress.CompressedXContent;
 import org.elasticsearch.common.settings.Settings;
 import org.elasticsearch.common.unit.TimeValue;
+import org.elasticsearch.common.util.concurrent.CountDown;
```
Comment (mikemccand) [inline core/src/test/java/org/elasticsearch/index/IndexServiceTests.java]: Not used?

File: core/src/test/java/org/elasticsearch/index/IndexServiceTests.java [hunk 2]
```diff
@@ -187,11 +189,18 @@ protected String getThreadPool() {
...
         };
+
+        BiFunction<AtomicReference<CountDownLatch>, CountDownLatch, CountDownLatch> swapAndReturn = (ref, newLatch) -> {
+            CountDownLatch downLatch = ref.get();
+            ref.set(newLatch);
+            return downLatch;
+        };
         latch.get().await();
-        latch.set(new CountDownLatch(1));
+        swapAndReturn.apply(latch, new CountDownLatch(1));
```
Comment (mikemccand) [inline core/src/test/java/org/elasticsearch/index/IndexServiceTests.java]: Can't we just keep the previous line here?  We don't use the returned (old latch) result?

File: core/src/test/java/org/elasticsearch/index/IndexServiceTests.java [hunk 3]
```diff
@@ -187,11 +189,18 @@ protected String getThreadPool() {
...
         assertEquals(1, count.get());
-        latch2.get().countDown();
-        latch2.set(new CountDownLatch(1));
+        // here we need to swap first before we let it go otherwise threads might be very fast and run that task twice due to
+        // random exception and the schedule interval is 1ms
+        swapAndReturn.apply(latch2, new CountDownLatch(1)).countDown();
```
Comment (mikemccand) [inline core/src/test/java/org/elasticsearch/index/IndexServiceTests.java]: If my previous comment was right :)  And this is now the only use of the new lambda, maybe don't use a lambda and inline the code here?

**Comment by mikemccand** [comment]
+1, sneaky concurrency horrors ... I left a couple comments on maybe avoiding a lambda here.

**Comment by s1monw** [comment]
@mikemccand pushed updates - less fancy same thing

**Comment by mikemccand** [comment]
+1, thanks @s1monw!

**Pull request merged**